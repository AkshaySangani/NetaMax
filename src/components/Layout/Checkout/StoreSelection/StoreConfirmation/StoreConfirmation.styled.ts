import styled from "styled-components";

import { IAbsoluteDivProps } from "../StoreList/StoreList.styled";

export const AbsoluteDiv = styled.div<IAbsoluteDivProps>`
  position: absolute;
  top: 70%;
  left: 0%;
  background-color: white;
  width: 100%;
  height: 30%;
  border-radius: 32px 32px 0 0;
  z-index: 1;
  transition: 0.3s;
  box-shadow: 0px -0.5px 2px rgba(0, 0, 0, 0.25);
  padding: 0 20px;
  display: flex;
  flex-flow: column;
`;
