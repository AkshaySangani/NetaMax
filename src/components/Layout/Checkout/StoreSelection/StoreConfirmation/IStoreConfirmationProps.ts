import { INearestStore } from "dataflows/Stores/INearestStore";

export interface IStoreConfirmationProps {
  /**
   * Selected store
   * @type {INearestStore}
   */
  store: INearestStore;

  /**
   * Clears the store selected
   * @type {() => void}
   */
  handleChooseAnotherStore: () => void;

  /**
   * Confirms the store
   * @type {() => void}
   */
  handleStoreConfirmation: () => void;

  /**
   * Google distance in minutes
   * @type {string}
   */
  googleDistance: string;
}
