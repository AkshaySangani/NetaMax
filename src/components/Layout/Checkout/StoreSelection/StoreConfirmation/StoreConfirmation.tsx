import { ReactElement } from "react";

import IconPersonWalking from "styled/icons/CheckoutFlow/PersonWalking";
import { roundDistance } from "utils/stringUtils";

import { Box, Button, Flex, Text } from "@chakra-ui/react";

import { IStoreConfirmationProps } from "./IStoreConfirmationProps";

/**
 * The StoreConfirmation component.
 * @param { IStoreConfirmationProps } props the component props.
 * @returns { ReactElement } The StoreConfirmation component.
 */
export const StoreConfirmation = (props: IStoreConfirmationProps): ReactElement => {
  const { name, companyAddress } = props.store;
  const { handleChooseAnotherStore, handleStoreConfirmation, googleDistance } = props;

  return (
    <Box
      position="absolute"
      height="35%"
      bottom="0"
      left="0"
      width="100%"
      backgroundColor="white"
      padding={{ base: "16px", md: "30px" }}
    >
      <Text fontSize="16px" fontWeight="700">
        {name}
      </Text>
      <Flex alignItems="center" marginTop="5px" color="#009DB5">
        <IconPersonWalking color="#009DB5" />
        <Text fontSize="14px">Aprox {roundDistance(googleDistance)}</Text>
      </Flex>
      <Text marginTop="5px" fontSize="14px" color="#686868">
        {companyAddress}
      </Text>
      <Flex
        direction="column"
        position="absolute"
        bottom="0"
        left="0"
        width="100%"
        padding={{ base: "16px", md: "30px" }}
      >
        <Button
          marginTop="15px"
          _hover={{ backgroundColor: "none" }}
          _disabled={{ backgroundColor: "netaGray.500" }}
          colorScheme="netaBlue"
          textColor="white"
          width="100%"
          borderRadius="xl"
          fontSize="18px"
          onClick={handleStoreConfirmation}
          height="48px"
        >
          Elegir tienda
        </Button>
        <Flex alignItems="center" justifyContent="center" marginTop="10px">
          <Text
            onClick={handleChooseAnotherStore}
            color="#3870FF"
            fontSize="14px"
            fontWeight="700"
            cursor="pointer"
            decoration="underline"
          >
            Elegir otra tiendita
          </Text>
        </Flex>
      </Flex>
    </Box>
  );
};
