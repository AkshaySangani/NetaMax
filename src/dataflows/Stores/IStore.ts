import { ILocation } from "./IStoreState";

export interface IStore {
  /**
   * The unique identifier for the item.
   * @type {string}
   */
  id: string;

  /**
   * The name identifies the item.
   * @type {string}
   */
  name: string;

  /**
   * Is a url to the store
   * @type {string}
   */
  url: string;

  /**
   * Is a host to the store
   * @type {string}
   */
  hosts: string;

  /**
   * Company name.
   * @type {string}
   */
  companyName: string;

  /**
   * Company name.
   * @type {string}
   */
  companyAddress: string;

  /**
   * Company phone number.
   * @type {string}
   */
  companyPhoneNumber: string;
}

export interface IResponse {
  stores: IStoreResponse[];
  distance: number;
}

export interface IStoreResponse {
  /**
   * The stores id
   * @type {string}
   */
  id: string;

  /**
   * The stores name
   * @type {string}
   */
  name: string;

  /**
   * The stores url
   * @type {string}
   */
  url: string;

  /**
   * The stores hosts
   * @type {string}
   */
  hosts: string;

  /**
   * The stores company name
   * @type {string}
   */
  companyName: string;

  /**
   * The stores company address
   * @type {string}
   */
  companyAddress: string;

  /**
   * The stores company phone number
   * @type {string}
   */
  companyPhoneNumber: string;

  /**
   * The stores distance in meters
   * @type {string}
   */
  storeDistance: string;

  /**
   * The stores distance in minutes
   * @type {string}
   */
  storeDistanceInMinutes: string;

  /**
   * The stores latitud
   * @type {string}
   */
  latitud: string;

  /**
   * The stores longitud
   * @type {string}
   */
  longitud: string;

  /**
   * Location coordinates of the store.
   * @type {ILocation}
   */
  location?: ILocation;
}
