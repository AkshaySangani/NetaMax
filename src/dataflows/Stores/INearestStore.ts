import { IStore } from "./IStore";
import { ILocation } from "./IStoreState";

export interface INearestStore extends IStore {
  /**
   * The distance in meters to the store
   * @type {string}
   */
  storeDistance: string;

  /**
   *  The distance in minutes to the store
   * @type {string}
   */
  storeDistanceInMinutes: string;

  /**
   * Location coordinates of the store.
   * @type {ILocation}
   */
  location: ILocation;
}
