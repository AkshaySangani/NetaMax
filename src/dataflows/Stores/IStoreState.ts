import { INearestStore } from "./INearestStore";
import { IStore } from "./IStore";

export interface IStoreState {
  /**
   * The array of stores
   * @type {IStore}
   **/
  store?: IStore;

  /**
   * Indicates if the products are loading.
   * @type {boolean}
   */
  isLoadingStore: boolean;

  /**
   * Indicates if the nearest stores are loading
   * @type {boolean}
   **/
  isLoadingNearestStores: boolean;

  /**
   * Indicates the nearest stores
   * @type {INearestStore[]}
   **/
  nearestStores: INearestStore[];

  /**
   * The error message
   * @type {string}
   */
  errorMessage?: string;

  /**
   * The error message
   * @type {ILocation}
   */
  userLocation?: ILocation;

  /**
   * selected store in map
   * @type {INearestStore}
   */
  selectedStore?: INearestStore;

  /**
   * Distance for the current locations search
   * @type {number}
   */
  circleDistance: number;
}

export interface ILocation {
  /**
   * Locations latitude
   * @type {number}
   */
  lat: number;

  /**
   * Locations longitude
   * @type {number}
   */
  lng: number;

  /**
   * Address name if possible
   * @type {string}
   */
  addressName?: string;
}
