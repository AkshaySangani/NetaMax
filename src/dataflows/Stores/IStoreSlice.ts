import { createSlice, PayloadAction } from "@reduxjs/toolkit";

import { IStore } from "./IStore";
import { IStoreState } from "./IStoreState";
import { getNearestStores, getStoreByName } from "./StoreThunks";

const initialState: IStoreState = {
  store: undefined,
  isLoadingStore: false,
  nearestStores: [],
  isLoadingNearestStores: false,
  circleDistance: 2500,
};

const storeSlice = createSlice({
  name: "store",
  initialState,
  reducers: {
    resetNearestStores: (state) => {
      state.nearestStores = [];
    },
    setStoreAsSelected: (state, action: PayloadAction<IStore>) => {
      state.store = action.payload;
    },
    setUserLocation: (state, action) => {
      state.userLocation = action.payload;
    },
    setSelectedStore: (state, action) => {
      state.selectedStore = action.payload;
    },
    clearSelectedStore: (state) => {
      state.selectedStore = undefined;
    },
    clearErrorMessage: (state) => {
      state.errorMessage = undefined;
    },
  },
  extraReducers: {
    [getStoreByName.pending.type]: (state) => {
      state.isLoadingStore = true;
    },
    [getStoreByName.fulfilled.type]: (state, action) => {
      state.store = action.payload;
      state.isLoadingStore = false;
    },
    [getNearestStores.pending.type]: (state) => {
      state.isLoadingNearestStores = true;
      state.nearestStores = [];
      state.circleDistance = 2500;
      state.errorMessage = undefined;
    },
    [getNearestStores.fulfilled.type]: (state, action) => {
      state.isLoadingNearestStores = false;
      state.nearestStores = action.payload.nearestStores;
      state.circleDistance = action.payload.distance;
      state.errorMessage = undefined;
    },
    [getNearestStores.rejected.type]: (state) => {
      state.isLoadingNearestStores = false;
      state.nearestStores = [];
      state.circleDistance = 2500;
      state.errorMessage =
        "No encontramos tienditas cercanas a la dirección añadida, intenta con otra.";
    },
  },
});

/**
 * Actions
 */
export const {
  resetNearestStores,
  setStoreAsSelected,
  setUserLocation,
  setSelectedStore,
  clearSelectedStore,
  clearErrorMessage,
} = storeSlice.actions;

/**
 * Reducers
 */
export const storeReducer = storeSlice.reducer;
