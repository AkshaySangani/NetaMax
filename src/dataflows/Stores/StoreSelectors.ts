import { RootState } from "state/store";

import { INearestStore } from "./INearestStore";
import { IStore } from "./IStore";
import { ILocation } from "./IStoreState";

/**
 * Selector to get the sales products.
 * @param {RootState} state the root state
 * @returns {IProduct[]} the sales products.
 */
export const selectStore = (state: RootState): IStore | undefined => state?.store?.store;

/**
 * Selector to get the isLoadingSalesSection state from the state.
 * @param {RootState} state the root state
 * @returns {boolean} the isLoadingSalesSection state
 */
export const selectIsLoadingStore = (state: RootState): boolean => state?.store?.isLoadingStore;

/**
 * Selector to get the nearest stores
 * @param {RootState} state the root state
 * @returns {Array} the nearest stores
 */
export const selectNearestStores = (state: RootState): INearestStore[] => state.store.nearestStores;

/**
 * Selector to get if is loading nearest stores
 * @param {RootState} state the root state
 * @returns {boolean} loading nearest stores
 */
export const selectLoadingNearestStores = (state: RootState): boolean =>
  state.store.isLoadingNearestStores;

/**
 * Gets the error message
 * @param {RootState} state the root state
 * @returns {string | undefined} the error message
 */
export const selectErrorMessage = (state: RootState): string | undefined =>
  state.store.errorMessage;

/**
 * Gets the user location
 * @param {RootState} state the root state
 * @returns {ILocation | undefined} the user location
 */
export const selectUserLocation = (state: RootState): ILocation | undefined =>
  state.store.userLocation;

/**
 * Gets the selected store
 * @param {RootState} state the root state
 * @returns {INearestStore | undefined} the selected store
 */
export const selectSelectedStore = (state: RootState): INearestStore | undefined =>
  state.store.selectedStore;

/**
 * Gets the circle distance
 * @param {RootState} state the root state
 * @returns {number} the error message
 */
export const selectCircleDistance = (state: RootState): number => state.store.circleDistance;
