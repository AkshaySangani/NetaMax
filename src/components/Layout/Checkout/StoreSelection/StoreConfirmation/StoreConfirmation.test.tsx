import { cleanup, render, screen } from "@testing-library/react";
import { INearestStore } from "dataflows/Stores/INearestStore";
import { IStoreConfirmationProps } from "./IStoreConfirmationProps";
import { StoreConfirmation } from "./StoreConfirmation";

describe("Render StoreConfirmation component", () => {
  let storeMocked: INearestStore;
  let mock: IStoreConfirmationProps;
  beforeEach(() => {
    storeMocked = {
      id: "2",
      name: "Super",
      storeDistance: "10.3",
      companyAddress: "Avenida insurgentes Sur 1886, Florida, Ciudad de México, CDMX, México.",
      url: "",
      companyName: "",
      hosts: "",
      companyPhoneNumber: "",
      latlong: {
        lat: 19.43215411,
        lng: -99.14388543,
      },
    };
    mock = {
      store: storeMocked,
      handleChooseAnotherStore: jest.fn(),
      handleStoreConfirmation: jest.fn(),
    };
  });

  afterEach(() => {
    cleanup();
  });

  test("StoreConfirmation should be rendered correctly", () => {
    const { container } = render(<StoreConfirmation {...mock} />);
    expect(container).toMatchSnapshot();
  });

  test("Store name should be rendered properly", () => {
    render(<StoreConfirmation {...mock} />);
    expect(screen.getByText(mock.store.name)).toBeInTheDocument();
  });

  test("Store company address should be rendered properly", () => {
    render(<StoreConfirmation {...mock} />);
    expect(screen.getByText(mock.store.companyAddress)).toBeInTheDocument();
  });

  test("'Elegir tienda' should be rendered properly", () => {
    render(<StoreConfirmation {...mock} />);
    expect(screen.getByText("Elegir tienda")).toBeInTheDocument();
  });

  test("'Elegir otra tiendita' should be rendered properly", () => {
    render(<StoreConfirmation {...mock} />);
    expect(screen.getByText("Elegir otra tiendita")).toBeInTheDocument();
  });
});
