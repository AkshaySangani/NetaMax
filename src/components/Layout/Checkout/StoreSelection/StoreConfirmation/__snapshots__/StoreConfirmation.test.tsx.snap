// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render StoreConfirmation component StoreConfirmation should be rendered correctly 1`] = `
<div>
  <div
    class="css-1c110c9"
  >
    <p
      class="chakra-text css-ils4e4"
    >
      Super
    </p>
    <div
      class="css-1h6ulav"
    >
      <svg
        fill="none"
        height="14"
        viewBox="0 0 501 689"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          clip-path="url(#clip0_6208_42124)"
        >
          <path
            d="M250.36 429.42C244.88 441.09 239.49 452.48 234.17 463.9C220.89 492.38 207.633 520.86 194.4 549.34C192.413 553.768 189.544 557.746 185.97 561.03C150.39 592.97 116.31 626.5 81.35 659.1C75.35 664.67 69.49 670.31 63.56 675.93C48.33 690.36 24.94 690.03 10.56 675.17C7.09417 671.611 4.37531 667.394 2.56357 662.768C0.751839 658.142 -0.116191 653.2 0.0105343 648.233C0.137259 643.267 1.25619 638.376 3.30149 633.848C5.34679 629.321 8.27713 625.248 11.92 621.87C50.2334 585.503 88.62 549.223 127.08 513.03C129.254 510.99 131.014 508.548 132.26 505.84C151.07 465.47 170.05 425.17 188.8 384.77C196.67 367.77 200.58 349.5 205.47 331.55C215.183 295.95 224.85 260.337 234.47 224.71C234.59 224.27 234.62 223.81 234.81 222.65C227.18 225.56 220.17 229.06 212.75 230.9C198.52 234.41 190.68 243.09 184.84 256.49C171.17 287.87 155.84 318.55 141.1 349.43C134.05 364.16 117.83 370.82 102.96 365.43C85.63 359.17 77.83 339.88 85.9 322.78C99.3733 294.207 112.91 265.67 126.51 237.17C131.51 226.66 136.44 216.12 141.45 205.62C146.3 195.48 153.51 188.17 164.4 184.06C198.4 171.22 232.313 158.07 266.14 144.61C275.38 140.91 284.08 136.04 294.14 134.71C320.34 131.24 344.7 145.48 355.32 169.79C360.53 181.72 359.69 194.26 360.77 206.59C363.01 232.13 364.72 257.73 366.42 283.32C366.65 286.82 368.11 288.02 371.15 289.01C407.29 300.883 443.397 312.84 479.47 324.88C492.66 329.3 500.68 341.88 499.91 355.88C499.462 362.324 496.984 368.461 492.832 373.411C488.681 378.36 483.068 381.868 476.8 383.43C470.606 384.896 464.118 384.528 458.13 382.37C417.93 369.003 377.71 355.69 337.47 342.43C331.47 340.43 325.25 339.09 319.84 335.71C317.77 334.42 317.6 335.84 317.27 337.04C313.75 350.04 310.35 363.04 306.71 376.04C306.03 378.47 307.32 379.32 308.71 380.46C339.23 404.68 369.57 429.12 400.33 453.03C414.04 463.69 418.74 477.7 417.15 493.94C413.39 532.38 408.97 570.76 404.81 609.16C403.14 624.55 401.63 639.97 399.81 655.34C397.38 675.49 378.38 690.15 358.21 687.71C337.75 685.22 323.36 666.47 325.65 645.87C330.65 601.19 335.07 556.46 340.41 511.87C341.35 503.98 339.33 499.69 333.26 495C306.79 474.53 280.75 453.51 254.54 432.69C253.31 431.65 251.99 430.67 250.36 429.42Z"
            fill="#009DB5"
          />
          <path
            d="M278.95 65.39C278.923 48.0342 285.793 31.3788 298.046 19.0876C310.3 6.7965 326.934 -0.123487 344.29 -0.150009C361.646 -0.176531 378.301 6.69257 390.592 18.9462C402.883 31.1998 409.804 47.8342 409.83 65.19C409.92 101.19 380.57 130.78 344.66 130.9C336.032 130.956 327.479 129.3 319.495 126.03C311.51 122.76 304.253 117.94 298.143 111.849C292.033 105.757 287.191 98.5148 283.896 90.5405C280.602 82.5661 278.921 74.0179 278.95 65.39V65.39Z"
            fill="#009DB5"
          />
        </g>
        <defs>
          <clippath
            id="clip0_6208_42124"
          >
            <rect
              fill="white"
              height="688.01"
              width="500.02"
            />
          </clippath>
        </defs>
      </svg>
      <p
        class="chakra-text css-1kuy7z7"
      >
        Aprox 
        NaN minutos
      </p>
    </div>
    <p
      class="chakra-text css-1kq5ykq"
    >
      Avenida insurgentes Sur 1886, Florida, Ciudad de México, CDMX, México.
    </p>
    <div
      class="css-hmr99h"
    >
      <button
        class="chakra-button css-lv84wk"
        type="button"
      >
        Elegir tienda
      </button>
      <div
        class="css-1yfye7s"
      >
        <p
          class="chakra-text css-9n5jqp"
        >
          Elegir otra tiendita
        </p>
      </div>
    </div>
  </div>
</div>
`;
